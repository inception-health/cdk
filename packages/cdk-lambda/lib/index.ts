export { Lambda } from "./lambda";
export type { LambdaProps } from "./lambda";

export { DeadLetterQueue } from "./dlq/queue";
export type { DeadLetterQueueProps } from "./dlq/queue";

export { StandardNetwork } from "./network/network";
export type { StandardNetworkProps } from "./network/network";
