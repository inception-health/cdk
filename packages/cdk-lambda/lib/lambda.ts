import {
  aws_iam,
  aws_kms,
  aws_lambda,
  aws_lambda_nodejs,
  aws_secretsmanager,
  aws_logs,
  Duration,
  Size,
} from "aws-cdk-lib";
import { Construct } from "constructs";

import { DeadLetterQueue } from "./dlq/queue";

/**
 * Interface for LambdaProps
 */
export interface LambdaProps {
  /**
   * Activate a Dead Letter Queue for the lambda function
   *
   * @defaultValue false
   */
  deadLetterQueue?: boolean;

  /**
   * Description of the lambda function
   */
  description?: string;

  /**
   * Encryption key to be used
   */
  encryption?: aws_kms.IKey;

  /**
   * Environment variables for the lambda function
   */
  environment?: { [key: string]: string };

  /**
   * File containing the lambda function
   */
  file: string;

  /**
   * Handler for the lambda function
   *
   * @defaultValue "handler"
   */
  handler?: string;

  /**
   * Log retention period in days
   *
   * @defaultValue 2 months
   */
  logRetention?: aws_logs.RetentionDays;

  /**
   * Logging level
   *
   * @defaultValue INFO
   */
  loggingLevel?: "INFO" | "DEBUG" | "ERROR";

  /**
   * Memory size of the Lambda function in MB.
   *
   * @defaultValue `128` (MB)
   */
  memorySize?: number;

  /**
   * Name of the lambda function
   */
  name: string;

  /**
   * Ephemeral storage for the Lambda function.
   *
   * @defaultValue 512 MB
   */
  storage?: Size;

  /**
   * Timeout of the Lambda function.
   *
   * @defaultValue 30 seconds
   */
  timeout?: Duration;
}

const defaultProps: Partial<LambdaProps> = {
  handler: "handler",
  logRetention: aws_logs.RetentionDays.TWO_MONTHS,
  timeout: Duration.seconds(30),
  memorySize: 128,
  loggingLevel: "INFO",
  storage: Size.mebibytes(512),
  deadLetterQueue: false,
};

export interface LambdaCDK {
  encryption: aws_kms.IKey | undefined;
  function: aws_lambda.Function;
  logGroup: aws_logs.LogGroup;
  role: aws_iam.Role;
  secret: aws_secretsmanager.ISecret | undefined;
}

/**
 * `Lambda` is a class that represents an AWS Lambda function in a CDK
 * application.
 *
 * By default, the log group is encrypted with the key provided in the
 * `encryption` prop. If no key is provided, the log group is encrypted with the
 * default key provided by AWS. The log retention is set to 1 year by default.
 *
 * @example
 * ```typescript
 * const myLambda = new Lambda({
 *   name: 'myLambdaFunction',
 *   handler: 'index.handler',
 *   secretName: 'mySecret',
 * });
 * ```
 *
 * @remarks
 * - **Logging**: The lambda function logs to CloudWatch Logs. The Lambda
 *   captures logs in JSON structured format without having to use your own
 *   logging libraries. JSON structured logs make it easier to search, filter,
 *   and analyze large volumes of log entries. For more information, read about
 *   the new [AWS Lambda advanced
 *   logging](https://aws.amazon.com/blogs/compute/introducing-advanced-logging-controls-for-aws-lambda-functions/).
 *
 */
export class Lambda extends Construct implements aws_iam.IGrantable {
  public readonly cdk: LambdaCDK;

  public readonly dlq: {
    processor: Lambda | undefined;
    queue: DeadLetterQueue | undefined;
  };

  private readonly props: LambdaProps;

  constructor(scope: Construct, id: string, props: LambdaProps) {
    super(scope, id);

    // Initialize the properties
    this.props = Object.assign({}, defaultProps, props);
    this.cdk = {} as LambdaCDK;
    this.dlq = {} as Lambda["dlq"];

    // Setup
    this.setupLogGroup();
    this.setupRole();
    this.setupDlq();
    this.setupFunction();
    this.setupEncryption();
    this.setupDlqProcessor();
  }

  // !Public
  public get grantPrincipal(): aws_iam.IPrincipal {
    return this.cdk.role;
  }

  public addEnvironment(key: string, value: string) {
    this.cdk.function.addEnvironment(key, value);
  }

  public grantInvoke(identity: aws_iam.IGrantable): aws_iam.Grant {
    return this.cdk.function.grantInvoke(identity);
  }

  // !Setup
  private setupLogGroup() {
    this.cdk.logGroup = new aws_logs.LogGroup(this, "log-group", {
      logGroupName: `/aws/lambda/${this.props.name}`,
      encryptionKey: this.props.encryption,
      retention: this.props.logRetention,
    });
  }

  private setupRole() {
    this.cdk.role = new aws_iam.Role(this, "role", {
      assumedBy: new aws_iam.ServicePrincipal("lambda.amazonaws.com"),
    });
    this.cdk.role.addManagedPolicy(
      aws_iam.ManagedPolicy.fromAwsManagedPolicyName(
        "service-role/AWSLambdaBasicExecutionRole",
      ),
    );
  }

  private setupDlq() {
    if (!this.props.deadLetterQueue) return;

    this.dlq.queue = new DeadLetterQueue(this, "dlq", {
      encryption: this.props.encryption,
      name: this.props.name,
    });
  }

  private setupFunction() {
    this.cdk.function = new aws_lambda_nodejs.NodejsFunction(this, "function", {
      // Default
      architecture: aws_lambda.Architecture.ARM_64,
      tracing: aws_lambda.Tracing.ACTIVE,
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      // Role
      role: this.cdk.role,
      // Performance
      timeout: this.props.timeout,
      memorySize: this.props.memorySize,
      ephemeralStorageSize: this.props.storage,
      // CloudWatch Lambda Insights is a monitoring and troubleshooting
      // solution for serverless applications running on AWS Lambda. The
      // solution collects, aggregates, and summarizes system-level metrics
      // including CPU time, memory, disk, and network. It also collects,
      // aggregates, and summarizes diagnostic information such as cold starts
      // and Lambda worker shutdowns to help you isolate issues with your Lambda
      // functions and resolve them quickly.
      // Ref: https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/Lambda-Insights.html
      insightsVersion: aws_lambda.LambdaInsightsVersion.VERSION_1_0_229_0,
      // Logs
      logGroup: this.cdk.logGroup,
      logFormat: aws_lambda.LogFormat.JSON,
      applicationLogLevel: aws_lambda.ApplicationLogLevel.INFO,
      systemLogLevel: aws_lambda.SystemLogLevel.INFO,
      // Customization
      functionName: this.props.name,
      description: this.props.description,
      // Environment and Encryption
      environmentEncryption: this.props.encryption,
      environment: this.props.environment,
      // Function
      entry: this.props.file,
      handler: this.props.handler,
      // Bundling
      bundling: {
        minify: true,
        sourceMap: false,
        externalModules: ["@aws-sdk/*"],
      },
    });
  }

  private setupEncryption() {
    if (!this.props.encryption) return;

    // Setup encryption
    this.cdk.encryption = this.props.encryption;
    this.cdk.encryption.grantDecrypt(this.cdk.function);
  }

  private setupDlqProcessor() {
    if (!this.props.deadLetterQueue) return;

    if (!this.dlq.queue) throw new Error("DLQ is undefined");

    this.dlq.processor = this.dlq.queue.attachRedrive(this);
  }
}
