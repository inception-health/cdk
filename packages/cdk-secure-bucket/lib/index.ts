export * from "./s3.secure";
export * from "./lifecycle";
