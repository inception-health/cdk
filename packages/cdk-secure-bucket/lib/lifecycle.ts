import { aws_s3, Duration } from "aws-cdk-lib";

// The lifecycle policy for items should be managed cohesively
export class LifecycleMode {
  static readonly NONE = new LifecycleMode();

  static readonly BASIC = new LifecycleMode({
    enabled: true,
    transitions: [
      {
        storageClass: aws_s3.StorageClass.INFREQUENT_ACCESS,
        transitionAfter: Duration.days(360),
      },
      {
        storageClass: aws_s3.StorageClass.GLACIER,
        transitionAfter: Duration.days(360 * 3),
      },
    ],
  });

  static readonly INTELLIGENT = new LifecycleMode({
    enabled: true,
    transitions: [
      {
        storageClass: aws_s3.StorageClass.INTELLIGENT_TIERING,
        transitionAfter: Duration.days(360),
      },
      {
        storageClass: aws_s3.StorageClass.GLACIER,
        transitionAfter: Duration.days(360 * 3),
      },
    ],
  });

  private constructor(public readonly rule?: aws_s3.LifecycleRule) {}
}
