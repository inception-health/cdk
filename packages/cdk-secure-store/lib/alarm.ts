import {
  Duration,
  aws_cloudwatch,
  aws_cloudwatch_actions,
  aws_dynamodb,
  aws_sns,
} from "aws-cdk-lib";
import { Construct } from "constructs";

import { SecureStore } from "./store";

export interface SecureStoreAlarmsProps {
  /**
   * The SNS topic to send DevOps notifications to. The SNS topic must be
   * created in the same account and region as the DynamoDB table.
   */
  devOpsTopicArn?: string;
  /**
   * Enable System Error Alarm
   *
   * @defaultValue true
   */
  systemErrorAlarm?: boolean;

  /**
   * Enable Throttle Alarm
   *
   * @defaultValue true
   */
  throttleAlarm?: boolean;
}

interface SecureStoreAlarmsCdk {
  table: aws_dynamodb.ITable;
  topic: aws_sns.ITopic | undefined;
  widgets: aws_cloudwatch.IWidget[];
}

const defaultProps: Partial<SecureStoreAlarmsProps> = {
  systemErrorAlarm: true,
  throttleAlarm: true,
};

/**
 * `SecureStoreAlarms` is a construct that creates alarms for monitoring a
 * secure store.
 *
 * This construct is responsible for creating CloudWatch alarms for various
 * metrics related to a secure store.
 *
 * The alarms created by this construct can be used to notify operators when
 * certain conditions are met.
 *
 * @remarks
 * We currently have two alarms:
 * - DynamoDB Table Reads/Writes Throttled: This alarm triggers when the
 *   DynamoDB table is throttled for reads or writes. This alarm is triggered
 *   when the DynamoDB table is throttled for 2% of the time in the last 30
 *   minutes.
 * - DynamoDB Errors > 0: This alarm triggers when there are any errors in the
 *   DynamoDB table. This alarm is triggered when there are any errors in the
 *   DynamoDB table in the last 30 minutes.
 *
 * @example
 * ```
 * // Creating a new instance of SecureStoreAlarms
 * const secureStoreAlarms = new SecureStoreAlarms(stack, 'MySecureStoreAlarms', {
 *   secureStore: mySecureStore,
 * });
 * ```
 *
 * @see
 * https://github.com/cdk-patterns/serverless/blob/main/the-cloudwatch-dashboard/README.md
 */
export class SecureStoreAlarms extends Construct {
  public readonly cdk: SecureStoreAlarmsCdk;

  // !Private Properties
  private readonly props: SecureStoreAlarmsProps;

  constructor(scope: SecureStore, id: string, props: SecureStoreAlarmsProps) {
    super(scope, id);

    // Initialize the properties
    this.props = { ...defaultProps, ...props };
    this.cdk = {} as SecureStoreAlarmsCdk;
    this.cdk.widgets = [];

    this.cdk.table = scope.cdk.table;

    if (this.props.devOpsTopicArn)
      this.cdk.topic = aws_sns.Topic.fromTopicArn(
        this,
        "DevOpsTopic",
        this.props.devOpsTopicArn,
      );

    // ==================================
    // System Errors
    // ==================================
    if (this.props.systemErrorAlarm) this.setupSystemError();

    // ==================================
    // Read/Write Throttles
    // ==================================
    if (this.props.throttleAlarm) this.setupThrottle();

    // ==================================
    // Cloudwatch Widgets
    // ==================================
    this.setupCloudwatchWidgets();
  }

  // !Private Methods
  private buildGraphWidget(
    widgetName: string,
    metrics: aws_cloudwatch.IMetric[],
    stacked = false,
  ): aws_cloudwatch.GraphWidget {
    const widget = new aws_cloudwatch.GraphWidget({
      title: widgetName,
      left: metrics,
      stacked: stacked,
      width: 8,
    });
    this.cdk.widgets.push(widget);
    return widget;
  }

  // !Private Setup
  private setupSystemError() {
    // ----------------------------------
    // Metric
    // ----------------------------------
    const metric = this.cdk.table.metricSystemErrorsForOperations({
      operations: [
        aws_dynamodb.Operation.GET_ITEM,
        aws_dynamodb.Operation.PUT_ITEM,
        aws_dynamodb.Operation.DELETE_ITEM,
        aws_dynamodb.Operation.UPDATE_ITEM,
        aws_dynamodb.Operation.TRANSACT_WRITE_ITEMS,
        aws_dynamodb.Operation.TRANSACT_GET_ITEMS,
        aws_dynamodb.Operation.BATCH_GET_ITEM,
        aws_dynamodb.Operation.BATCH_WRITE_ITEM,
        aws_dynamodb.Operation.QUERY,
        aws_dynamodb.Operation.SCAN,
      ],
      period: Duration.minutes(5),
    });

    // ----------------------------------
    // Alarm
    // ----------------------------------
    const alarm = new aws_cloudwatch.Alarm(this, "alarm-errors", {
      alarmName: `DynamoDB ${this.cdk.table.tableName} System Errors`,
      // Metric
      metric: metric,
      // Configuration
      threshold: 0,
      evaluationPeriods: 6,
      datapointsToAlarm: 1,
      treatMissingData: aws_cloudwatch.TreatMissingData.NOT_BREACHING,
    });

    // ----------------------------------
    // Actions
    // ----------------------------------
    // If a topic is defined in the CDK stack, an SNS action is added to the
    // alarm. This means that when the alarm state changes, a message will be
    // published to the SNS topic.
    if (this.cdk.topic)
      alarm.addAlarmAction(
        new aws_cloudwatch_actions.SnsAction(this.cdk.topic),
      );
  }

  private setupThrottle() {
    // ----------------------------------
    // Metric
    // ----------------------------------
    // The purpose of this metric is to monitor the throttling events of
    // DynamoDB. It combines the `ReadThrottleEvents` and `WriteThrottleEvents` into
    // a single metric. The statistic used is `'sum'`, which means it will add up
    // all the throttle events in the specified period. The period is set to 5
    // minutes, meaning the metric data points are emitted every 5 minutes.
    const metric = new aws_cloudwatch.MathExpression({
      expression: "m1 + m2",
      label: `DynamoDB ${this.cdk.table.tableName} Reads/Writes Throttled`,
      usingMetrics: {
        m1: this.cdk.table.metric("ReadThrottleEvents", { statistic: "sum" }),
        m2: this.cdk.table.metric("WriteThrottleEvents", { statistic: "sum" }),
      },
      period: Duration.minutes(5),
    });

    // ----------------------------------
    // Alarm
    // ----------------------------------
    // The purpose of this alarm is to trigger when there are any DynamoDB
    // throttling events. The alarm is configured to trigger when the metric
    // value is greater than 0. The evaluationPeriods is set to 6, meaning the
    // alarm will look at the last 6 periods (or 30 minutes) of data. The
    // datapointsToAlarm is set to 1, meaning the alarm will trigger if any
    // single data point in the evaluation period breaches the threshold. The
    // treatMissingData is set to NOT_BREACHING, meaning if there are any
    // periods of missing data, they will not count as breaching the threshold.
    const alarm = new aws_cloudwatch.Alarm(this, "alarm-throttle", {
      alarmName: `DynamoDB ${this.cdk.table.tableName} Throttled`,
      // Metric
      metric: metric,
      // Configuration
      threshold: 0,
      evaluationPeriods: 6,
      datapointsToAlarm: 1,
      treatMissingData: aws_cloudwatch.TreatMissingData.NOT_BREACHING,
    });

    // ----------------------------------
    // Actions
    // ----------------------------------
    // If a topic is defined in the CDK stack, an SNS action is added to the
    // alarm. This means that when the alarm state changes, a message will be
    // published to the SNS topic.
    if (this.cdk.topic)
      alarm.addAlarmAction(
        new aws_cloudwatch_actions.SnsAction(this.cdk.topic),
      );
  }

  private setupCloudwatchWidgets() {
    // ----------------------------------
    // DynamoDB Latency
    // ----------------------------------
    this.buildGraphWidget(
      `DynamoDB ${this.cdk.table.tableName} Latency`,
      ["GetItem", "UpdateItem", "PutItem", "DeleteItem", "Query"].map((op) => {
        return this.cdk.table.metricSuccessfulRequestLatency({
          dimensionsMap: {
            TableName: this.cdk.table.tableName,
            Operation: op,
          },
        });
      }),
      true,
    );

    // ----------------------------------
    // DynamoDB Consumed Read/Write Units
    // ----------------------------------
    this.buildGraphWidget(
      `DynamoDB ${this.cdk.table.tableName} Consumed Read/Write Units`,
      [
        this.cdk.table.metricConsumedReadCapacityUnits(),
        this.cdk.table.metricConsumedWriteCapacityUnits(),
      ],
      true,
    );

    // ----------------------------------
    // DynamoDB Throttles
    // ----------------------------------
    this.buildGraphWidget(
      `DynamoDB ${this.cdk.table.tableName} Throttles`,
      [
        this.cdk.table.metric("ReadThrottleEvents", { statistic: "sum" }),
        this.cdk.table.metric("WriteThrottleEvents", { statistic: "sum" }),
      ],
      true,
    );
  }
}
