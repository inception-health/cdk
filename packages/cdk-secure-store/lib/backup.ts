import { aws_backup, aws_dynamodb, aws_kms } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface SecureStoreBackupProps {
  backupPlanId?: string;
  encryptionKey: aws_kms.IKey;
  table: aws_dynamodb.ITableV2;
  tableName: string;
}

export interface SecureStoreBackupResources {
  backupPlan: aws_backup.IBackupPlan | undefined;
  backupVault: aws_backup.BackupVault | undefined;
}

export class SecureStoreBackup extends Construct {
  public readonly cdk: SecureStoreBackupResources;

  private readonly props: SecureStoreBackupProps;

  constructor(scope: Construct, id: string, props: SecureStoreBackupProps) {
    super(scope, id);

    // Initialize the properties
    this.props = props;
    this.cdk = {} as SecureStoreBackupResources;

    // Create the backup resources
    if (this.props.backupPlanId) {
      this.cdk.backupPlan = aws_backup.BackupPlan.fromBackupPlanId(
        this,
        "backupPlan",
        this.props.backupPlanId,
      );
    } else {
      // Create a backup vault
      this.cdk.backupVault = new aws_backup.BackupVault(this, "backupVault", {
        backupVaultName: `secure-store-${this.props.tableName}-backup-vault`,
        encryptionKey: this.props.encryptionKey,
        blockRecoveryPointDeletion: true,
        // TODO: Review if we want to make lock configuration by default.
        // lockConfiguration
      });

      // Create the backup plan
      this.cdk.backupPlan =
        aws_backup.BackupPlan.dailyWeeklyMonthly7YearRetention(
          this,
          `secure-store-${this.props.tableName}-backup-plan`,
          this.cdk.backupVault,
        );

      (this.cdk.backupPlan as aws_backup.BackupPlan).addSelection(
        `secure-store-${this.props.tableName}-backup-plan-selection`,
        {
          resources: [
            aws_backup.BackupResource.fromDynamoDbTable(this.props.table),
          ],
        },
      );
    }
  }
}
