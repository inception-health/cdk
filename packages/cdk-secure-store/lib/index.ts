export type { SecureStoreProps } from "./store";
export { SecureStore } from "./store";
