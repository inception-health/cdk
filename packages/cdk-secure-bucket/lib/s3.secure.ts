import { aws_kms, aws_s3 } from "aws-cdk-lib";
import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";

import { LifecycleMode } from "./lifecycle";

export interface SecureStoreProps
  extends Pick<aws_s3.BucketProps, "bucketName"> {
  /**
   * External KMS key to use for bucket encryption.
   *
   * The `encryption` property must be specified.
   *
   */
  encryptionKey: string;

  /**
   * Type of Bucket lifecycle to use.
   *
   * - Basic: simple basic lifecycle
   * - Intelligent: intelligent lifecycle
   * - None: no lifecycle (not recommended)
   *
   *@defaultValue BASIC
   */
  lifecycleMode?: LifecycleMode;

  /**
   * Bucket removal policy.
   *
   *@defaultValue RemovalPolicy.RETAIN
   */
  removalPolicy?: cdk.RemovalPolicy;
}

const defaultProps = {
  lifecycleMode: LifecycleMode.BASIC,
  removalPolicy: cdk.RemovalPolicy.RETAIN,
};

/**
 * A secure S3 Bucket that  follows Inception Health HIPAA policies
 *
 * References:
 * - https://docs.aws.amazon.com/AmazonS3/latest/userguide/security-best-practices.html
 */
export class SecureBucket extends aws_s3.Bucket {
  private readonly props: SecureStoreProps;
  public readonly kmsKey: aws_kms.IKey;

  constructor(scope: Construct, id: string, props: SecureStoreProps) {
    const customKmsKey = aws_kms.Key.fromKeyArn(
      scope,
      `${id}-kms`,
      props.encryptionKey,
    );
    const config = Object.assign({}, defaultProps, props);

    // validation
    if (!config || !config.lifecycleMode)
      throw new Error("Construct requires default lifecycle");

    super(scope, id, {
      bucketName: props.bucketName,
      // ================================================
      // Lifecycle Management
      // ================================================
      lifecycleRules: config.lifecycleMode.rule
        ? [config.lifecycleMode.rule]
        : undefined,
      removalPolicy: config.removalPolicy,

      //autoDeleteObjects: false, // this is applicable only when removalPolicy` to be set to `RemovalPolicy.DESTROY`

      // ================================================
      // Inventory Management
      // ================================================
      // References:
      // - https://docs.aws.amazon.com/AmazonS3/latest/userguide/storage-inventory.html
      // inventories: ,

      // ================================================
      // Preventative Security Best Practices
      // ================================================
      // Step 1- Use Amazon S3 block public access.
      blockPublicAccess: aws_s3.BlockPublicAccess.BLOCK_ALL,
      // Step 2- Encryption
      encryption: aws_s3.BucketEncryption.KMS,
      // S3 Bucket Keys decrease the request traffic from Amazon S3 to AWS Key Management Service (AWS KMS) and reduce the cost of SSE-KMS
      // https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucket-key.html
      bucketKeyEnabled: true,
      encryptionKey: customKmsKey,
      // Step 3 - Enforce encryption of data in transit
      enforceSSL: true,
      // Step 4 - Versioning
      // Versioning is a means of keeping multiple variants of an object in the same bucket.
      // You can use versioning to preserve, retrieve, and restore every version of every
      // object stored in your Amazon S3 bucket. With versioning, you can easily recover
      // from both unintended user actions and application failures.
      // TODO: Also consider implementing on-going detective controls using
      // the s3-bucket-versioning-enabled managed AWS Config rule.
      versioned: true,

      // Step 5 - Cross-region replication
      // TODO: Consider Amazon S3 cross-region replication

      // Step 6 - VPC endpoints for Amazon S3 access
      //
      // TODO: Consider VPC endpoints for Amazon S3 access VPC endpoints for
      // Amazon S3 provide multiple ways to control access to your Amazon S3
      // data:
      // - You can control the requests, users, or groups that are allowed
      //   through a specific VPC endpoint.
      // - You can control which VPCs or VPC endpoints have access to your S3
      //   buckets by using S3 bucket policies.
      // - You can help prevent data exfiltration by using a VPC that does not
      //   have an internet gateway.

      // ================================================
      // Monitoring and Auditing Best Practices
      // ================================================
      // Step 1 - Identify and audit all your Amazon S3 buckets

      // Step 3 - Enable Amazon S3 server access logging
      // TODO: should we build a custom s3 access log destination
      serverAccessLogsPrefix: "accesslogs",
    });

    // ================================================
    // Monitoring and Auditing Best Practices
    // ================================================
    // Step 2 - Implement monitoring using AWS monitoring tools
    // Monitoring is an important part of maintaining the reliability,
    // security, availability, and performance of Amazon S3 and your AWS
    // solutions. AWS provides several tools and services to help you
    // monitor Amazon S3 and your other AWS services. For example, you can
    // monitor CloudWatch metrics for Amazon S3, particularly PutRequests,
    // GetRequests, 4xxErrors, and DeleteRequests. For more information, see
    // Monitoring metrics with Amazon CloudWatch and, Monitoring Amazon S3.
    // For a second example, see Example: Amazon S3 Bucket Activity. This
    // example describes how to create an Amazon CloudWatch alarm that is
    // triggered when an Amazon S3 API call is made to PUT or DELETE bucket
    // policy, bucket lifecycle, or bucket replication, or to PUT a bucket ACL.
    // const putRequestMetric = this.mm({ id: 'PutRequests' });
    // putRequestMetric.

    // Step 4 - Use AWS CloudTrail

    // Step 5 - Enable AWS Config

    // Step 6 - Consider using Amazon Macie with Amazon S3

    // Step 7 - Monitor AWS security advisories

    // ================================================
    // Props
    // ================================================
    this.props = config;
    this.kmsKey = customKmsKey;
  }
}
