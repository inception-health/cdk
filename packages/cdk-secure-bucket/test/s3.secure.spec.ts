import * as crypto from "crypto";

import { Stack } from "aws-cdk-lib";
import * as cdk from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";

import { LifecycleMode } from "../lib/lifecycle";
import { SecureBucket } from "../lib/s3.secure";

describe("Secure S3 Store", () => {
  it("should conform to a basic snapshot", async () => {
    // GIVEN
    const stack = new Stack();
    const kmsArn = "arn:aws:kms:us-east-1:123454:key/general";
    new SecureBucket(stack, "test-secure-store", { encryptionKey: kmsArn });

    // THEN
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });

  it("should allow configuration of the bucket name", async () => {
    // GIVEN
    const stack = new Stack();
    const customBucketName = `custom-${crypto.randomUUID()}`;
    const customKmsArn = "arn:aws:kms:us-east-1:123454:key/general";
    new SecureBucket(stack, "test-secure-store", {
      encryptionKey: customKmsArn,
      bucketName: customBucketName,
    });

    // THEN
    Template.fromStack(stack).hasResourceProperties("AWS::S3::Bucket", {
      BucketName: customBucketName,
    });
  });

  it("should have versioning enabled by default", async () => {
    // GIVEN
    const stack = new Stack();
    const customBucketName = `custom-${crypto.randomUUID()}`;
    const customKmsArn = "arn:aws:kms:us-east-1:123454:key/general";
    new SecureBucket(stack, "test-secure-store", {
      encryptionKey: customKmsArn,
      bucketName: customBucketName,
    });

    // THEN
    Template.fromStack(stack).hasResourceProperties("AWS::S3::Bucket", {
      VersioningConfiguration: {
        Status: "Enabled",
      },
    });
  });

  it("should replace type of lifecycle to intelligent", async () => {
    // GIVEN
    const stack = new Stack();
    const customBucketName = `custom-${crypto.randomUUID()}`;
    const customKmsArn = "arn:aws:kms:us-east-1:123454:key/general";
    new SecureBucket(stack, "test-secure-store", {
      encryptionKey: customKmsArn,
      bucketName: customBucketName,
      lifecycleMode: LifecycleMode.INTELLIGENT,
    });

    // THEN
    Template.fromStack(stack).hasResourceProperties("AWS::S3::Bucket", {
      LifecycleConfiguration: {
        Rules: [
          {
            Status: "Enabled",
            Transitions: [
              {
                StorageClass: "INTELLIGENT_TIERING",
                TransitionInDays: 360,
              },
              {
                StorageClass: "GLACIER",
                TransitionInDays: 1080,
              },
            ],
          },
        ],
      },
    });
  });

  it("should replace type of lifecycle to NONE", async () => {
    // GIVEN
    const stack = new Stack();
    const customBucketName = `custom-${crypto.randomUUID()}`;
    const customKmsArn = "arn:aws:kms:us-east-1:123454:key/general";
    new SecureBucket(stack, "test-secure-store", {
      encryptionKey: customKmsArn,
      bucketName: customBucketName,
      lifecycleMode: LifecycleMode.NONE,
    });

    // THEN
    Template.fromStack(stack).hasResourceProperties("AWS::S3::Bucket", {
      LifecycleConfiguration: Match.absent(),
    });
  });

  it("should configure removal policy to DESTROY", async () => {
    // GIVEN
    const stack = new Stack();
    const customBucketName = `custom-${crypto.randomUUID()}`;
    const customKmsArn = "arn:aws:kms:us-east-1:123454:key/general";
    new SecureBucket(stack, "test-secure-store", {
      encryptionKey: customKmsArn,
      bucketName: customBucketName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    // THEN
    Template.fromStack(stack).hasResource("AWS::S3::Bucket", {
      DeletionPolicy: "Delete",
    });
  });
});
