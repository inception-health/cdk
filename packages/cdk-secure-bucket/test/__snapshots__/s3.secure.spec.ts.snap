// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Secure S3 Store should conform to a basic snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testsecurestore7F781630": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessControl": "LogDeliveryWrite",
        "BucketEncryption": {
          "ServerSideEncryptionConfiguration": [
            {
              "BucketKeyEnabled": true,
              "ServerSideEncryptionByDefault": {
                "KMSMasterKeyID": "arn:aws:kms:us-east-1:123454:key/general",
                "SSEAlgorithm": "aws:kms",
              },
            },
          ],
        },
        "LifecycleConfiguration": {
          "Rules": [
            {
              "Status": "Enabled",
              "Transitions": [
                {
                  "StorageClass": "STANDARD_IA",
                  "TransitionInDays": 360,
                },
                {
                  "StorageClass": "GLACIER",
                  "TransitionInDays": 1080,
                },
              ],
            },
          ],
        },
        "LoggingConfiguration": {
          "LogFilePrefix": "accesslogs",
        },
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "ObjectWriter",
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testsecurestorePolicy4033C827": {
      "Properties": {
        "Bucket": {
          "Ref": "testsecurestore7F781630",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "testsecurestore7F781630",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "testsecurestore7F781630",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
